### DISCUSSION TITLE
Function hooked to finalize_distribution_options gets called for sdist+wheel+editable install

### DISCUSSION THREAD

**Discussion by smasihemami** [pypa/setuptools#4349] [Q&A]
I have a python package that infers version of another package from git and is Pep518 enabled, meaning target packages should have this package in their pyproject.toml under build requires section for it to work. It has a function that hooks to setuptools finalize_distribution_options and does the job. What I need to do is to make sure that it only gets called when creating an sdist, and afterwards store version information somewhere and re-use again when creating a wheel or editable install. I need to ensure that this versioning package is a build time dependency only when creating an sdist to sort of speed up the pip install of a target package which has this versioning package as a build time dependency. Is there a way to tell setuptools in pyproject.toml that a dependency is only needed for creation of sdist, and not others?

**Comment by abravalheri** [comment]
Can you use `setuptools-scm` for that? I believe that they already have this part figured out.

Regarding calling the plugin only for `sdist`: there is currently no support for that. The function will be called in all build hooks defined for PEP 517 and 660. Moreover, it conceptually doesn't work: there are build hooks that need to extract the complete metadata but run before `sdist`.

It is up to the function implementation to always return the same thing, but you can check for the existence of caching.

The easy thing that you can do is to check if the `PKG-INFO` file exists and get the info from there before attempting anything else.

But again, I think it is just easier to use `setuptools-scm` because they already have solved most edge cases.

**Reply by smasihemami** [reply]
Thanks for your answer. Will for sure try setuptools-scm, but the challenge there is that I then have to update pyproject.toml for all the consumers of this versioning package.
Could you please advise further on how to check for existence of PKG-INFO file given a setuptools.Distribution object in my hook function?

**Reply by abravalheri** [reply]
> Thanks for your answer. Will for sure try setuptools-scm, but the challenge there is that I then have to update pyproject.toml for all the consumers of this versioning package.

They will have to add something to `[build-system] requires` in `pyproject.toml` anyway right? So adding your package there or `setuptools-scm` is kind of the same thing isn't it?

Now the question is: do you need your plugin just to set `version` or is it something else that your package will do? Because you can also add `setuptools-scm` as a dependency of your package if you need something else, and use set the `.order` attribute in your hook function so that it loads after `setuptools-scm`.

> Could you please advise further on how to check for existence of PKG-INFO file given a setuptools.Distribution object in my hook function?

This is something you have to try out and see what works best. Very likely that `pathlib.Path('PKG-INFO').is_file()` is enough of a check. Then you should be able to read the version from that file (one of the many metadata attributes in the file)
